File: kafkaStreamsModelServerKVStore/src/main/java/com/lightbend/standard/modelserver/ModelProcessor.java
package com.lightbend.standard.modelserver;

import com.lightbend.configuration.kafka.ApplicationKafkaParameters;
import com.lightbend.model.ModelWithDescriptor;
import com.lightbend.model.ModelServingInfo;
import com.lightbend.standard.modelserver.store.StoreState;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;
import org.apache.kafka.streams.state.KeyValueStore;

import java.util.Objects;
import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 */
public class ModelProcessor extends AbstractProcessor<byte[], Optional<ModelWithDescriptor>> {

    private KeyValueStore<Integer, StoreState> modelStore;

    @Override
    public void process(byte[] key, Optional<ModelWithDescriptor> modelWithDescriptor) {

        StoreState state = modelStore.get(ApplicationKafkaParameters.STORE_ID);
        if(state == null)
            state = new StoreState();
        state.setNewModel(modelWithDescriptor.get().getModel());
        state.setNewServingInfo(new ModelServingInfo(modelWithDescriptor.get().getDescriptor().getName(),
                    modelWithDescriptor.get().getDescriptor().getDescription(), 0));
        modelStore.put(ApplicationKafkaParameters.STORE_ID, state);
        return;
    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = (KeyValueStore<Integer, StoreState>) context.getStateStore(ApplicationKafkaParameters.STORE_NAME);
        Objects.requireNonNull(modelStore, "State store can't be null");
    }
}

File: kafkaStreamsModelServerCustomStore/src/main/scala/com/lightbend/scala/custom/modelserver/ModelProcessor.scala
package com.lightbend.scala.custom.modelserver

import java.util.Objects

import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext, ProcessorSupplier}
import com.lightbend.java.configuration.kafka.ApplicationKafkaParameters
import com.lightbend.scala.modelServer.model.{ModelToServeStats, ModelWithDescriptor}
import com.lightbend.scala.custom.store.ModelStateStore

import scala.util.Try

class ModelProcessor extends AbstractProcessor[Array[Byte], Try[ModelWithDescriptor]] with ProcessorSupplier[Array[Byte], Try[ModelWithDescriptor]]{

  private var modelStore: ModelStateStore = null

  import ApplicationKafkaParameters._
  override def process (key: Array[Byte], modelWithDescriptor: Try[ModelWithDescriptor]): Unit = {

    modelStore.state.newModel = Some(modelWithDescriptor.get.model)
    modelStore.state.newState = Some(ModelToServeStats(modelWithDescriptor.get.descriptor))
  }

  override def init(context: ProcessorContext): Unit = {
    modelStore = context.getStateStore(STORE_NAME).asInstanceOf[ModelStateStore];
    Objects.requireNonNull(modelStore, "State store can't be null")
  }

  override def get() = new ModelProcessor()
}
